crates/ty_ide/src/completion.rs

```diff
@@ -4034,66 +4017,144 @@ def f(x: T):
     x.<CURSOR>
 ",
         );
-        test.assert_completions_include("on_a_and_b");
-        test.assert_completions_do_not_include("only_on_a");
-        test.assert_completions_do_not_include("only_on_b");
+        let test = builder.build();
+
+        test.contains("on_a_and_b");
+        test.not_contains("only_on_a");
+        test.not_contains("only_on_b");
     }
 
     #[test]
     fn typevar_without_bounds_or_constraints() {
-        let test = cursor_test(
+        let test = completion_test_builder(
             "\
 def f[T](x: T):
     x.<CURSOR>
 ",
         );
-        test.assert_completions_include("__repr__");
+        test.build().contains("__repr__");
     }
 
-    // NOTE: The methods below are getting somewhat ridiculous.
-    // We should refactor this by converting to using a builder
-    // to set different modes. ---AG
+    /// A way to create a simple single-file (named `main.py`) completion test
+    /// builder.
+    ///
+    /// Use cases that require multiple files with a `<CURSOR>` marker
+    /// in a file other than `main.py` can use `CursorTest::builder()`
+    /// and then `CursorTestBuilder::completion_test_builder()`.
+    fn completion_test_builder(source: &str) -> CompletionTestBuilder {
+        CursorTest::builder()
+            .source("main.py", source)
+            .completion_test_builder()
+    }
 
-    impl CursorTest {
-        /// Returns all completions except for builtins.
-        fn completions_without_builtins(&self) -> String {
-            self.completions_if(|c| !c.builtin)
+    /// A builder for executing a completion test.
+    ///
+    /// This mostly owns the responsibility for generating snapshots
+    /// of completions from a cursor position in source code. Most of
+    /// the options involve some kind of filtering or adjustment to
+    /// apply to the snapshots, depending on what one wants to test.
+    struct CompletionTestBuilder {
+        cursor_test: CursorTest,
+        settings: CompletionSettings,
+        skip_builtins: bool,
+        type_signatures: bool,
+        predicate: Option<Box<dyn Fn(&Completion) -> bool>>,
+    }
+
+    impl CompletionTestBuilder {
+        /// Returns completions based on this configuration.
+        fn build(&self) -> CompletionTest<'_> {
+            let original = completion(
+                &self.cursor_test.db,
+                &self.settings,
+                self.cursor_test.cursor.file,
+                self.cursor_test.cursor.offset,
+            );
+            let filtered = original
```

Comment:
Nit: You'll probably role your eyes given that this is a test. Please feel free to disregard it ;)

Should we use `original.drain` when building `filtered` and a) only store the filtered out suggestions in `CompletionTest` (instead of `original`) or b) only store a bool saying if any suggestions were discarded.